src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPathExtensions.cs

```diff
@@ -52,7 +62,77 @@ public static IEdmNavigationSource NavigationSource(this ODataPath path)
         /// <returns>True if the resource if a resource set or collection of primitive or complex types. False otherwise.</returns>
         public static bool IsCollection(this ODataPath path)
         {
-            return path.LastSegment.TranslateWith(new IsCollectionTranslator());
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
+            return (bool)path.LastSegment?.TranslateWith(IsCollectionTranslator.Instance);

```

Comment:
If `LastSegment` can be null, then there's the possibility of this statement evaluating to `(bool)null`. Won't that throw an exception?